### PR TITLE
JDK-8315951: Open source several Swing HTMLEditorKit related tests

### PR DESCRIPTION
**Pull request by honkar-jdk**
Following tests are open sourced as part of this PR.

1. java/awt/event/PaintEvent/RepaintTest/RepaintTest.java
2. javax/swing/text/html/HTMLEditorKit/4214848/bug4214848.java
3. javax/swing/text/html/HTMLEditorKit/4230197/bug4230197.java
4. javax/swing/text/html/HTMLEditorKit/4238223/bug4238223.java

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315951](https://bugs.openjdk.org/browse/JDK-8315951): Open source several Swing HTMLEditorKit related tests (**Bug** - P4)

### Reviewers
 * [Damon Nguyen](https://openjdk.org/census#dnguyen) (@DamonGuy - Committer) ⚠️ Review applies to [a8984390](https://git.openjdk.org/jdk/pull/15751/files/a898439070dcad61beef31fae50be3c25b12a255)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15751/head:pull/15751` \
`$ git checkout pull/15751`

Update a local copy of the PR: \
`$ git checkout pull/15751` \
`$ git pull https://git.openjdk.org/jdk.git pull/15751/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15751`

View PR using the GUI difftool: \
`$ git pr show -t 15751`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15751.diff">https://git.openjdk.org/jdk/pull/15751.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15751#issuecomment-1719975580)

### CODE REVIEW

**Review by DamonGuy — APPROVED** [review]
Looks fine to me.

File: test/jdk/javax/swing/text/html/HTMLEditorKit/bug4230197.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        kit.insertHTML(doc, doc.getLength(), "<sup>0</sup>", 0, 0, HTML.Tag.SUP);
+        kit.insertHTML(doc, doc.getLength(), "<b>0</b>", 0, 0, HTML.Tag.B);
+        kit.insertHTML(doc, doc.getLength(), "<i>0</i>", 0, 0, HTML.Tag.I);
+        kit.insertHTML(doc, doc.getLength(), "<code>0</code>", 0, 0, HTML.Tag.CODE);
+        kit.write(sw, doc, 0, doc.getLength());
+
+        String au = sw.toString().toLowerCase();
+        if ((!au.contains("<sub>0</sub>")) || (!au.contains("<sup>0</sup>"))
+                || (!au.contains("<code>0</code>")) || (!au.contains("<b>0</b>"))
+                || (!au.contains("<i>0</i>"))) {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HTMLEditorKit/bug4230197.java:51]: I'd put each condition on its own line for easier parsing:
```suggestion
        if ((!au.contains("<sub>0</sub>"))
                || (!au.contains("<sup>0</sup>"))
                || (!au.contains("<code>0</code>"))
                || (!au.contains("<b>0</b>"))
                || (!au.contains("<i>0</i>"))) {
```

File: test/jdk/javax/swing/text/html/HTMLEditorKit/bug4238223.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        parser.parse(testHTML);
+    }
+
+    static class TestCallback extends HTMLEditorKit.ParserCallback {
+        String commentData = "comment";
+        int commentIndex = 65;
+
+        public void handleComment(char[] data, int pos) {
+            if (! (new String(data)).equals(commentData) ||
+                    pos != commentIndex) {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HTMLEditorKit/bug4238223.java:53]:
```suggestion
            if (!(new String(data)).equals(commentData)
                    || pos != commentIndex) {
```
There are no spaces after the `!` operator in the following `if` statements. I would also suggest wrapping before the operator.

File: test/jdk/javax/swing/text/html/HTMLEditorKit/bug4238223.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            }
+        }
+
+        HTML.Tag[] endTags = {HTML.Tag.TITLE, HTML.Tag.HEAD,
+                HTML.Tag.BODY, HTML.Tag.HTML};
+        int[] endTagPositions = {23, 31, 79, 86};
+        int endTagIndex = 0;
+        public void handleEndTag(HTML.Tag tag, int pos) {
+            if (! tag.equals(endTags[endTagIndex]) ||
+                    pos != endTagPositions[endTagIndex]) {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/HTMLEditorKit/bug4238223.java:65]:
```suggestion
            if (!tag.equals(endTags[endTagIndex])
                    || pos != endTagPositions[endTagIndex]) {
```
No space after `!`, wrap before the operator.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: test/jdk/java/awt/event/PaintEvent/RepaintTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            int count = counter.getCount();
+            robot.delay(300);
+
+            EventQueue.invokeAndWait(() -> panel.repaint());
+            robot.waitForIdle();
+            robot.delay(1000);
+
+            if (counter.getCount() == count) {
+                throw new RuntimeException("Failed");
+            }
```
[the remark this answers is not in the document]
Comment (honkar-jdk) [reply test/jdk/java/awt/event/PaintEvent/RepaintTest.java]: Updated

File: test/jdk/java/awt/event/PaintEvent/RepaintTest.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+            robot.waitForIdle();
+            robot.delay(1000);
+
+            EventQueue.invokeAndWait(() -> panel.setLocation(panel.getX() + 10,
+                                                             panel.getY() + 10));
+            robot.waitForIdle();
+            robot.delay(500);
+
+            int count = counter.getCount();
+            robot.delay(300);
```
[the remark this answers is not in the document]
Comment (honkar-jdk) [reply test/jdk/java/awt/event/PaintEvent/RepaintTest.java]: Delay removed.

File: test/jdk/java/awt/event/PaintEvent/RepaintTest.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * @test
+ * @bug 4394287
+ * @key headful
+ * @summary Paint pending on heavyweight component move
+ */
+
+// Subclass of Frame, creates two buttons: one that moves the Frame by
+// invoking setBounds, and the other that invokes repaint directly on
+// the Frame to repaint a Component. The Component displays an integer
+// that increments everytime paint is invoked on it.
```
[the remark this answers is not in the document]
Comment (honkar-jdk) [reply test/jdk/java/awt/event/PaintEvent/RepaintTest.java]: Removed comment

**Comment by honkar-jdk** [comment]
Review changes updated. PR ready for re-review.

**Review by aivanov-jdk — APPROVED** [review]

**Pull request closed without merging**